atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflowsTest.java

```diff
@@ -0,0 +1,214 @@
+        Consumer<List<CrossCellInconsistency>> violationConsumer = mock(Consumer.class);
+        WorkflowHistory history = transientRowsWorkflow.run();
+        transactionStore.readWrite(txn -> txn.delete(TABLE_NAME, ImmutableWorkloadCell.of(ITERATION_COUNT - 1, 1)));
+        invariant.accept(history, violationConsumer);
+        verify(violationConsumer)
+                .accept(ImmutableList.of(CrossCellInconsistency.builder()
+                        .putInconsistentValues(
+                                TableAndWorkloadCell.of(
+                                        TABLE_NAME,
+                                        ImmutableWorkloadCell.of(ITERATION_COUNT - 1, TransientRowsWorkflows.COLUMN)),
+                                Optional.empty())
+                        .putInconsistentValues(
+                                TableAndWorkloadCell.of(
+                                        TABLE_NAME,
+                                        ImmutableWorkloadCell.of(
+                                                TransientRowsWorkflows.SUMMARY_ROW, ITERATION_COUNT - 1)),
+                                Optional.of(0))
+                        .build()));
+        verifyNoMoreInteractions(violationConsumer);
+    }
+
+    @Test
+    public void invariantReportsViolationsFromFinalStateWhenExtraCellsArePresent() {
+        Consumer<List<CrossCellInconsistency>> violationConsumer = mock(Consumer.class);
+        WorkflowHistory history = transientRowsWorkflow.run();
+        transactionStore.readWrite(txn -> txn.write(
+                TABLE_NAME,
+                ImmutableWorkloadCell.of(ITERATION_COUNT - 2, TransientRowsWorkflows.COLUMN),
+                TransientRowsWorkflows.VALUE));
+        invariant.accept(history, violationConsumer);
+        verify(violationConsumer)
+                .accept(ImmutableList.of(CrossCellInconsistency.builder()
+                        .putInconsistentValues(
+                                TableAndWorkloadCell.of(
+                                        TABLE_NAME,
+                                        ImmutableWorkloadCell.of(ITERATION_COUNT - 2, TransientRowsWorkflows.COLUMN)),
+                                Optional.of(0))
+                        .putInconsistentValues(
+                                TableAndWorkloadCell.of(
+                                        TABLE_NAME,
+                                        ImmutableWorkloadCell.of(
+                                                TransientRowsWorkflows.SUMMARY_ROW, ITERATION_COUNT - 2)),
+                                Optional.empty())
+                        .build()));
+        verifyNoMoreInteractions(violationConsumer);
+    }
+
+    @SuppressWarnings("unchecked") // Mocking of a known type
+    @Test
+    public void invariantReportsViolationsFromTransactionHistory() {
+        Consumer<List<CrossCellInconsistency>> violationConsumer = mock(Consumer.class);
```

Comment:
Any reason to not just use an `AtomicReference::set`?

Reply:
I did think about that, but wanted to verify the consumer is only invoked once (and it felt like the instrumentation of managing a list of lists to check that we don't call it more than once was overkill). What do you think?

Reply:
maybe a half-way compromise would be to spy `AtomicReference`?

Reply:
I think you updated it everywhere but here :D

Reply:
you're right! fixed